Replace libuv with libevent

jagerman opened the pull request:
Builds on top of #12.

jagerman committed:
Remove debug counter

jagerman committed:
Replace libuv with libevent

jagerman committed:
speedtest: Add -3 option to disable xor checksum

tewinget commented on include/quic/endpoint.hpp:
```diff
@@ -108,17 +111,21 @@ namespace oxen::quic
         }
 
-        std::shared_ptr<uvw::loop> get_loop();
+        const std::shared_ptr<event_base>& get_loop() { return net.loop(); }
```
`const std::shared_ptr<event_base>& get_loop() const { return net.loop(); }`

jagerman replied:
This is returning a non-const event_base pointer (just wrapped in a const lvalue reference to the shared_ptr that holds it); the event_base it is returning, however, is non-const.

tewinget commented on include/quic/endpoint.hpp:
```diff
@@ -108,17 +111,21 @@ namespace oxen::quic
-        std::shared_ptr<uvw::loop> get_loop();
+        const std::shared_ptr<event_base>& get_loop() { return net.loop(); }
+
+        const std::unique_ptr<UDPSocket>& get_socket() { return socket; }
```
^ see above

jagerman replied:
^ see above

jagerman committed:
Re-add CLI11 exclusion to format.sh

jagerman committed:
Add options to alltests to control log output

tewinget commented on src/connection.cpp:
```diff
@@ -566,29 +539,15 @@ namespace oxen::quic
-        auto ep = stream->conn.endpoint();
-
-        if (ep)
+        if (uint64_t app_err_code = context->stream_open_cb ? context->stream_open_cb(*stream) : 0; app_err_code != 0)
```
I believe the `return NGTCP2_ERR_CALLBACK_FAILURE` below is connection-fatal.  Commenting to remind us to consider whether we want a stream cb failure to be connection-fatal.

tewinget replied:
Since `app_err_code` is ours to define, we may want some which are and some which are not connection-fatal.

jagerman replied:
Fixed.

tewinget commented on src/connection.cpp:
```diff
@@ -749,7 +711,7 @@ namespace oxen::quic
         ngtcp2_settings_default(&settings);
 
-        settings.initial_ts = get_timestamp();
+        settings.initial_ts = get_timestamp().count();
```
in `on_io_ready` you changed `get_timestamp` to `get_time`, would that change be appropriate here, or this change there instead?

jagerman replied:
`get_time` returns a steady_clock::time_point, this returns the .time_since_epoch() of that time_point.  (The main difference is that it now returns typesafe nanoseconds instead of a bare integer).

jagerman committed:
Make stream open callback error code return non-fatal

Previously when the stream-opened callback returned an error we were
returning a fatal error back to ngtcp2, which is a fatal error and will
kill the connection.  This changes it to immediately close the stream
with the given app error code, but otherwise keep the connection alive.
